engine/pkg/election/elector.go

```diff
@@ -0,0 +1,357 @@
+			zap.String("address", e.config.Address),
+		)
+	}
+
+	if err := s.Update(ctx, record); err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record)
+	return nil
+}
+
+func (e *Elector) release(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, defaultReleaseTimeout)
+	defer cancel()
+
+	s := e.config.Storage
+
+	record, err := s.Get(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record.Clone())
+
+	for i, m := range record.Members {
+		if m.ID == e.config.ID {
+			record.Members = append(record.Members[:i], record.Members[i+1:]...)
+			break
+		}
+	}
+	if record.LeaderID == e.config.ID {
+		record.LeaderID = ""
+	}
+
+	if err := s.Update(ctx, record); err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record)
+	return nil
+}
+
+func (e *Elector) setObservedRecord(record *Record) {
+	e.observeLock.Lock()
+	defer e.observeLock.Unlock()
+
+	// Remove members that are not in the new record.
+	for id := range e.observedRenews {
+		if _, ok := record.FindMember(id); !ok {
+			delete(e.observedRenews, id)
+		}
+	}
+
+	// Update observedRenews for members in the new record.
+	for _, m := range record.Members {
+		oldMember, ok := e.observedRecord.FindMember(m.ID)
+		// If the member is not in the old record, or the RenewTime is
+		// changed, update the local observedRenews.
+		if !ok || !oldMember.RenewTime.Equal(m.RenewTime) {
+			e.observedRenews[m.ID] = time.Now()
+		}
+	}
+
+	// New leader is elected.
+	if record.LeaderID != "" && record.LeaderID != e.observedRecord.LeaderID {
+		leader, ok := record.FindMember(record.LeaderID)
+		if ok {
+			log.Info(
+				"new leader elected",
+				zap.String("leader-id", leader.ID),
+				zap.String("leader-name", leader.Name),
+				zap.String("leader-address", leader.Address),
+			)
+		}
+	}
+
+	e.observedRecord = *record
+}
+
+func (e *Elector) isLeaseExpired(memberID string) bool {
+	e.observeLock.RLock()
+	defer e.observeLock.RUnlock()
+
+	return e.isLeaseExpiredLocked(memberID)
+}
+
+func (e *Elector) isLeaseExpiredLocked(memberID string) bool {
+	member, ok := e.observedRecord.FindMember(memberID)
+	if !ok {
+		return true
+	}
+	renewTime := e.observedRenews[memberID]
+	return renewTime.Add(member.LeaseDuration).Before(time.Now())
+}
+
+// IsLeader returns true if the current member is the leader
+// and its lease is still valid.
+func (e *Elector) IsLeader() bool {
+	e.observeLock.RLock()
+	defer e.observeLock.RUnlock()
+
+	if e.isLeaseExpiredLocked(e.config.ID) {
+		return false
+	}
+	return e.observedRecord.LeaderID == e.config.ID
+}
+
+// GetLeader returns the last observed leader whose lease is still valid.
+func (e *Elector) GetLeader() (*Member, bool) {
+	e.observeLock.RLock()
+	defer e.observeLock.RUnlock()
+
+	for _, m := range e.observedRecord.Members {
+		if e.isLeaseExpiredLocked(m.ID) {
+			continue
+		}
+		if m.ID == e.observedRecord.LeaderID {
+			return m.Clone(), true
+		}
```

Comment:
Is it better to check `m.ID == e.observedRecord.LeaderID` first, and break loop immediately if lease of leader id is expired.

Reply:
Optimized in https://github.com/pingcap/tiflow/pull/6949/commits/69242db0678f279014e9043e3454e8a82cdd19d4.